src/ows/src/main/java/org/geoserver/ows/Dispatcher.java

```diff
@@ -1590,51 +1563,83 @@ Object parseRequestXML(Object requestBean, BufferedReader input, Request request
+     * <li>{@link Request#setRequest: The xml root element name, assuming it matches
+     * the request name, might be overriten later while parting the request's query
+     * string key-value pairs
+     * <li>{@link Request#setService service}: matching the xml root element's
+     * {@code service} attribute, or {@code null}
+     * <li>{@link Request#setVersion version}: matching the xml root element's
+     * {@code version} attribute, or {@code null}
+     * <li>{@link Request#setOutputFormat outputFormat}: matching the xml root
+     * element's {@code outputFormat} attribute, or {@code null}
+     * </ul>
      *
-     * @param input {@link BufferedReader} containing a valid OWS XML request body
+     * @param req The request to set properties to based on the xml request body's
+     *            root element
      * @return a {@link Map} containing the parsed parameters.
      * @throws Exception if there was an error reading the input.
      */
-    public static Map readOpPost(BufferedReader input) throws Exception {
+    public Request readOpPost(Request req) throws Exception {
+        char[] buff = new char[XML_LOOKAHEAD];
+        {
+            // Read into buff and use that for the XML stream reader, using req.getInput() directly
+            // can mess up the BufferedReader's state depending on the implementation
+            @SuppressWarnings("PMD.CloseResource")
+            BufferedReader input = req.getInput();
+            input.mark(XML_LOOKAHEAD);
+            input.read(buff);
+            input.reset();
+        }
         // create stream parser
-        XmlPullParserFactory factory = XmlPullParserFactory.newInstance();
-        factory.setNamespaceAware(true);
-        factory.setValidating(false);
-
-        // parse root element
-        XmlPullParser parser = factory.newPullParser();
-        parser.setInput(input);
-        parser.nextTag();
-
-        Map<String, String> map = new HashMap<>();
-        map.put("request", parser.getName());
-        map.put("namespace", parser.getNamespace());
-
-        for (int i = 0; i < parser.getAttributeCount(); i++) {
-            String attName = parser.getAttributeName(i);
-
-            if ("service".equals(attName)) {
-                map.put("service", parser.getAttributeValue(i));
-            }
-
-            if ("version".equals(parser.getAttributeName(i))) {
-                map.put("version", parser.getAttributeValue(i));
-            }
-
-            if ("outputFormat".equals(attName)) {
-                map.put("outputFormat", parser.getAttributeValue(i));
+        XMLInputFactory factory = XMLInputFactory.newFactory();
+        // disable DTDs
+        factory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+        // disable external entities
+        factory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false);
+        XMLStreamReader parser = factory.createXMLStreamReader(new CharArrayReader(buff));
```

Comment:
Shouldn't this reader be closed?

Reply:
no real need to, as it's running of a `CharArrayReader(char[])`, but still, won't hurt, so closing it

Reply:
https://github.com/geoserver/geoserver/pull/4804/commits/86998b5ed3b342f44e2387a50b28077b37d3b16e#diff-09c0707df546d28c44971393a12fb1318901427a2f7957becdc5e69ecd315ee7R1595-R1619